libs/gretty/src/main/groovy/org/akhikhl/gretty/ProjectUtils.groovy

```diff
@@ -84,20 +84,37 @@ final class ProjectUtils {
   }
 
   private static void addProjectClassPathWithDependencies(Set<URL> urls, Project proj, String dependencyConfigName) {
-    if(proj.sourceSets.main.output.files.any { new File(it, 'META-INF/web-fragment.xml').exists() })
-      urls.add proj.jar.archivePath.toURI().toURL()
-    else
-      urls.addAll proj.sourceSets.main.output.files.collect { it.toURI().toURL() }
     def dependencyConfig = proj.configurations.findByName(dependencyConfigName)
     if(dependencyConfig) {
-      def files = dependencyConfig.fileCollection { !(it instanceof ProjectDependency) }
+      Set<File> dependencyProjectArchives = new LinkedHashSet<>()
+
+      def collectProjectDependencies
+      collectProjectDependencies= { Project project ->
+        def outputFiles = project.sourceSets.main.output.files
+        if(outputFiles.any { new File(it, 'META-INF/web-fragment.xml').exists() }) {
+          urls.add(project.jar.archivePath.toURI().toURL())
+        } else {
+          urls.addAll(outputFiles.collect { it.toURI().toURL() })
+          dependencyProjectArchives.add(project.jar.archivePath)
+        }
+
+        def depConfig = project.configurations.findByName(dependencyConfigName)
+        if (depConfig) {
+          depConfig.allDependencies.withType(ProjectDependency).each {
+            collectProjectDependencies(it.dependencyProject)
+          }
+        }
+      }
+
+      collectProjectDependencies(proj)
```

Comment:
The logic here is trying to resolve every sub-project and add its output folder as the classpath. It will also collect its output jar file, and remove the jar files from the classpath to avoid duplication at the end.

Reply:
Thank you so much for giving an example showing the actual impact of the change. It helps a lot 👍